### PR TITLE
Feature - Add fpgaBindSVA() to match OPAE

### PR DESCRIPTION
**Pull request by michael-adler**
### Description
Allocates a PASID with the same interface as OPAE. Does not yet emulate PCIe ATS/PRS with PASID.

### Collateral (docs, reports, design examples, case IDs):

- [ ] Document Update Required? (Specify FIM/AFU/Scripts)

### CODE REVIEW

File: ase/api/src/buffer.c [hunk 1]
```diff
@@ -532,3 +533,40 @@ fpga_result __FPGA_API__ ase_fpgaGetIOAddress(fpga_handle handle, uint64_t wsid,
 	}
 	return result;
 }
+
+/*
+ * A single PASID is associated with the process, independent of the number
+ * of FPGA ports connected or the open/close sequence.
+ */
+static int process_pasid;
```
Comment (tswhison) [inline ase/api/src/buffer.c:541]: May want to init to -1.  0 is a valid fd.

File: ase/api/src/buffer.c [hunk 2]
```diff
@@ -532,3 +533,40 @@ fpga_result __FPGA_API__ ase_fpgaGetIOAddress(fpga_handle handle, uint64_t wsid,
...
+
+fpga_result __FPGA_API__ ase_fpgaBindSVA(fpga_handle handle, uint32_t *pasid)
+{
+	struct _fpga_handle *_handle = (struct _fpga_handle *)handle;
+	fpga_result result = FPGA_OK;
+	int err;
+
+	if (!pasid)
```
Comment (tswhison) [inline ase/api/src/buffer.c:549]: The top-level doc allows pasid to be NULL.  Doc update or change the behavior here, I think.

File: ase/api/src/buffer.c [hunk 3]
```diff
@@ -532,3 +533,40 @@ fpga_result __FPGA_API__ ase_fpgaGetIOAddress(fpga_handle handle, uint64_t wsid,
...
+		return FPGA_INVALID_PARAM;
+
+	result = handle_check_and_lock(_handle);
+	if (result)
+		return result;
+
+	/* Allocate a PASID if the process doesn't have one yet */
+	if (!process_pasid) {
```
Comment (tswhison) [inline ase/api/src/buffer.c:557]:
```suggestion
	if (process_pasid < 0) {
```

File: ase/api/src/buffer.c [hunk 4]
```diff
@@ -532,3 +533,40 @@ fpga_result __FPGA_API__ ase_fpgaGetIOAddress(fpga_handle handle, uint64_t wsid,
...
+		/* PASID is 20 bits, non-zero */
+		process_pasid = rand() & 0xfffff;
```
Comment (tswhison) [inline ase/api/src/buffer.c:559]: Do you need something like 1 + rand() % 0xffffff so that it is non-zero?

File: ase/api/src/types_int.h [hunk 5]
```diff
@@ -95,7 +95,8 @@ struct _fpga_handle {
 	void *umsg_virt;	    // umsg Virtual Memory pointer
 	uint64_t umsg_size;	    // umsg Virtual Memory Size
 	uint64_t *umsg_iova;	    // umsg IOVA from driver
-	bool fpgaMMIO_is_mapped;  // is MMIO mapped?
+	int pasid;
+	bool fpgaMMIO_is_mapped;    // is MMIO mapped?
```
Comment (tswhison) [inline ase/api/src/types_int.h:99]: Is this being set?  I didn't notice it.

**Review by tswhison — COMMENTED** [review]
👍

File: ase/api/src/buffer.c (hunk 1 above)
Comment (michael-adler) [reply ase/api/src/buffer.c:541]: process_pasid is guaranteed non-zero when valid. It isn't an fd.

File: ase/api/src/buffer.c (hunk 3 above)
Comment (michael-adler) [reply ase/api/src/buffer.c:557]: process_pasid is guaranteed non-zero when valid. It isn't an fd.

File: ase/api/src/buffer.c (hunk 4 above)
Comment (michael-adler) [reply ase/api/src/buffer.c:559]: See line 561

File: ase/api/src/types_int.h (hunk 5 above)
Comment (michael-adler) [reply ase/api/src/types_int.h:99]: Yes. Toward the end of ase_fpgaBindSVA().

File: ase/api/src/buffer.c (hunk 4 above)
Comment (tswhison) [reply ase/api/src/buffer.c:556]: doh, I see.

**Review by tswhison — APPROVED** [review]

**Pull request merged**